File: 2024/1/1.py
def solvePart1(fileName):
  lines = readFile(fileName).split('\n')
  a, b = getLists(lines)
  a.sort()
  b.sort()

  # assumes both lists are the same length
  total = 0
  for i, a in enumerate(a):
    total += abs(a - b[i])
  return total

def solvePart2(fileName):
  lines = readFile(fileName).split('\n')
  a, b = getLists(lines)

  similarity = 0
  for num in a:
    similarity += getSimilarity(num, timesInList(num, b))
  return similarity

def timesInList(num, list):
  total = 0
  for x in list:
    if x == num:
      total += 1
  return total

def getSimilarity(num, count):
   return num * count

def getLists(lines):
  a = []
  b = []
  for line in lines:
     pair = line.split('   ')
     a.append(int(pair[0]))
     b.append(int(pair[1]))
  return a, b

def readFile(fileName):
   with open(fileName) as f:
      return f.read()

if __name__ == "__main__":
    print('Part 1:', solvePart1('input.txt'))
    print('Part 2:', solvePart2('input.txt'))

File: 2024/1/1.rs
use std::fs;

fn main() {
  let input = fetch_input();
  let lines: Vec<_> = input.lines().collect();
  let part1 = part_one(lines.clone());
  let part2 = part_two(lines.clone());
  println!("Part 1: {part1}");
  println!("Part 2: {part2}");
}

fn part_one(lines: Vec<&str>) -> i32 {
  let (mut a, mut b) = get_lists(lines);
  a.sort();
  b.sort();
  
  let mut total = 0;
  for (i, num) in a.iter().enumerate() {
    total += i32::abs(num - b[i])
  }
  return total;
}

fn part_two(lines: Vec<&str>) -> i32 {
  let (a, b) = get_lists(lines);

  let mut similarity = 0;
  for num in a {
    similarity += num * times_in_list(num, b.clone());
  }
  return similarity;
}

fn get_lists(lines: Vec<&str>) -> (Vec<i32>, Vec<i32>) {
  let mut a = Vec::new();
  let mut b = Vec::new();

  for line in lines.into_iter() {
    let args: Vec<_> = line.split("   ").collect();
    a.push(args[0].parse::<i32>().unwrap());
    b.push(args[1].parse::<i32>().unwrap());
  }

  return (a, b);
}

fn times_in_list(num: i32, list: Vec<i32>) -> i32 {
  let mut total = 0;
  for x in list {
    if x == num {
      total += 1;
    }
  }
  return total;
}

fn fetch_input() -> String {
  let contents = fs::read_to_string("./input.txt")
    .expect("Input file not found.");
  return contents
}